### ISSUE TITLE
Eclipse and other SWT based applications doesn't properly load file chooser dialog

### ISSUE THREAD

**Issue by ongun-kanat** [lah7/gtk3-classic#17]
I am using the latest version of gtk3-mushrooms and love it. However, I encountered this problem that makes Eclipse unusable using gtk3-mushrooms package. The contents of the file dialog are not loaded.

![screenshot_20190203_060526](https://user-images.githubusercontent.com/3601965/52172903-b738f480-2779-11e9-9ab9-b3530dbdcedb.png)

![screenshot_20190203_060544](https://user-images.githubusercontent.com/3601965/52172906-c15af300-2779-11e9-885b-23cf40027ed9.png)

Problem persists in Adwaita theme too:

![screenshot_20190203_060832](https://user-images.githubusercontent.com/3601965/52172932-3e866800-277a-11e9-9c9a-7aa8975fc800.png)

An error is shown in the terminal when this happens:
```
(Eclipse:3883): Gtk-CRITICAL **: 06:05:23.075: Error building template class 'GtkFileChooserDialog' for an instance of type 'GtkFileChooserDialog': .:2:794 Invalid object type 'GtkFileChooserWidget'
```

**Comment by TomaszGasior** [comment]
Thanks for your activity.

First, please try to install bare gtk3 package from official repository and then try to reproduce the problem with non-patched gtk3.

Unfortunately I am not able to reproduce your problem. I am not using Eclipse, maybe I am doing something wrong. I did `yay -S eclipse` and then installed `eclipse-cpp` package. Next, I started app from terminal and went to File -> Open File. File selection dialog was shown properly.

Please, give me more information. Eclipse version (package name?), Exact procedure to reproduce your bug.

**Comment by ongun-kanat** [comment]
Similarly I installed `eclipse-common` and `eclipse-java` using pacman. It should be the same as yours (2018-12) since I also use Arch, I suppose. I also downloded the binary distribution of Eclipse. It behaves the same as well as  `jadx` a SWT based java decompiler IDE. I just launch eclipse from the terminal like 

`$ eclipse` and then click on the File -> Open File as you said.

Using `gtk3` package,  it behaves normally:

![screenshot_20190203_155703](https://user-images.githubusercontent.com/3601965/52178316-2be94e80-27cd-11e9-9eaa-17ee5630dbe7.png)

I also discovered running as `LD_PRELOAD=/usr/lib/libgtk-3.so eclipse` forcing libgtk-3 to be preloaded works and properly displays the file dialog using `gtk3-mushrooms`.

**Comment by TomaszGasior** [comment]
Unfortunately, still I am not able to reproduce your issue. I tried in clean Arch Linux with KDE and eclipse-java (from repo and downloaded from Eclipse website). On my computer, file chooser dialog is shown properly with gtk3-mushrooms. 

Please try to reproduce this bug with clean installation of Eclipse. Try to remove (move temporarily) your Eclipse configuration files  or, even better, run Eclipse from newly created user account on your Arch. Maybe there is something wrong between gtk3-mushrooms and your Eclipse configuration.

**Comment by ongun-kanat** [comment]
Okay I created a `test_xdg` directory and set it as my home using environment variables.

```
export HOME=/home/ongun/test_xdg
export XDG_CONFIG_HOME=/home/ongun/test_xdg/config
export XDG_DATA_HOME=/home/ongun/test_xdg/local
export XDG_CACHE_HOME=/home/ongun/test_xdg/cache
```
On top of that I removed .eclipse directory in my own homedir if eclipse can manage to find it via /etc/passwd.

Moreover I started with a nearly empty environment.

```
DISPLAY=:0
EDITOR=vim
HOME=/home/ongun/test_xdg
LANG=en_US.UTF-8
LANGUAGE=en_US
LESS_TERMCAP_se=
PATH=/usr/bin
PWD=/home/ongun/test_xdg
SHLVL=0
_=/usr/bin/env
XAUTHORITY=/tmp/xauth-1000-_0
XDG_CACHE_HOME=/home/ongun/test_xdg/cache
XDG_CONFIG_HOME=/home/ongun/test_xdg/config
XDG_DATA_HOME=/home/ongun/test_xdg/local
```

And then run eclipse. The result is the same.

![screenshot_20190204_201710](https://user-images.githubusercontent.com/3601965/52231347-025a2100-28ba-11e9-963f-1a30eb13dc92.png)

Maybe this is about Java version? Which version of Java do you have? I tried with both Java 8 OpenJDK and Java 11 OpenJDK?

**Comment by TomaszGasior** [comment]
Still I can't reproduce. I tried with `java-10-openjdk`, `java-11-openjdk`, `java-8-openjdk/jre`. No issue, file chooser dialog is shown properly. Hm, interesting…

**Comment by ongun-kanat** [comment]
This gets weird. Either we stumble upon a race condition or for whatever reason SWT cannot not correctly load gtk3. Because `LD_PRELOAD`ing libgtk-3.so works :/

**Comment by ongun-kanat** [comment]
After a long debug session. This is what I've found.

Somehow `gtk_file_chooser_widget_get_type` function which is created by the part/macro below in `gtkfilechooserwidget.c` is not called when I start the application. This function registers the `GtkFileChooserWidget` type into GLib type database.

The definition is in line 614, the macro expansion defines a function called `gtk_file_chooser_widget_get_type`:

```c
G_DEFINE_TYPE_WITH_CODE (GtkFileChooserWidget, gtk_file_chooser_widget, GTK_TYPE_BOX,
                         G_ADD_PRIVATE (GtkFileChooserWidget)
                         G_IMPLEMENT_INTERFACE (GTK_TYPE_FILE_CHOOSER,
                                                gtk_file_chooser_widget_iface_init)
                         G_IMPLEMENT_INTERFACE (GTK_TYPE_FILE_CHOOSER_EMBED,
                                                gtk_file_chooser_embed_default_iface_init));
```

This function must be called by `_gtk_builder_resolve_type_lazily`, defined in line 404 of `gtkbuilder.c`

```c
static GType
_gtk_builder_resolve_type_lazily (const gchar *name)
{
  static GModule *module = NULL;
  GTypeGetFunc func;
  gchar *symbol;
  GType gtype = G_TYPE_INVALID;

  if (!module)
    module = g_module_open (NULL, 0);

  symbol = type_name_mangle (name);

  if (g_module_symbol (module, symbol, (gpointer)&func))
    gtype = func ();

  g_free (symbol);

  return gtype;
}
```

However `g_module_symbol` fails for some reason and cannot find symbol hence it returns `FALSE` and the `func ()` which would be the pointer to `gtk_file_chooser_widget_get_type` isn't executed. The function returns `G_TYPE_INVALID` and the error message in the first comment is displayed.

I have a clue right now. Maybe, somehow libgtk-3.so isn't completely loaded into memory and force loading with `LD_PRELOAD` forces all of the libgtk-3.so to load and `g_module_symbol` works normally so my file dialogs are shown correctly.

**Comment by ongun-kanat** [comment]
Added a print line to output the `g_module_error`. It returns:  
`java: undefined symbol: gtk_file_chooser_widget_get_type`

I might have found the root cause:

https://mail.gnome.org/archives/gtk-list/2015-February/msg00080.html

It's unfortunately unanswered :/.

**Comment by TomaszGasior** [comment]
Please try to compile gtk3-mushrooms without ...gtk3-mushrooms patchset. To do it, follow:
`git clone https://github.com/TomaszGasior/gtk3-mushrooms.git`
`cd gtk3-mushrooms`
open PKGBUILD file and around line 140, inside `prepare()` function, comment out `__patch_gtk_code` call like here: https://i.imgur.com/YDok1r6.png , then
`makepkg -si`
This will compile and install gtk3-mushrooms (without patches). Then, try to reproduce error. Maybe there is something wrong in package configuration.

**Comment by ongun-kanat** [comment]
After all those research I tried compiling `gtk3` itself. Guess what happened? It also failed to load File dialog widget!

After fiddling many `makepkg` options and comparing. I eliminated the sources of confusion find the cause. It's the f*kin locale. GTK is the one of the core libraries of Linux distribution and they fail to implement multiple locale support correctly. They failed the notorious [The Turkey test](http://www.moserware.com/2008/02/does-your-code-pass-turkey-test.html) .

I am from Turkey, and although most (important) of my LC_* environment variables are "en_US.UTF-8" some of them like LC_COLLATE are "tr_TR.UTF-8". Due to whatever "ignorant developer" reason, the behavior of produced binary is completely different. It's so pathetic that I didn't expect the cause of the problem is simply gloabally unaware code considering the library is GTK. However, here we are :P

Using my software English and avoiding most of the problems and using simple tricks like LC_NUMBERS, LC_COLLATE I can make some software behave sane enough. This went straight home and made me remember the old times when I used my computer in Turkish and 75% of the FOSS software was failed to compile or run correctly.

Thanks for your time and attention. I guess I will fill a bug and even recommend using tr_TR.UTF-8 to reproducible builds project.

**Comment by TomaszGasior** [comment]
Thank you for your investigation!

**Comment by ongun-kanat** [comment]
FYI They fixed it for gtk-3.24 in https://gitlab.gnome.org/GNOME/gtk/merge_requests/577 and it's merged to gtk-3.24 branch. It would be nice to have the same patch in your repo.

**Comment by TomaszGasior** [comment]
Thank you. I will wait until new GTK release.
